### PR TITLE
fix: Load new state from session/local storage when available after a silent renew

### PR DESCRIPTION
**Pull request by Powersource**
Fixes https://github.com/bjerkio/oidc-react/issues/361

...I think. When I edited directly in node_modules (it complained when i npm linked because of duplicate react) it worked but this code is slightly different.

### CODE REVIEW

**Comment by Powersource** [comment]
Would love some help on why those tests are failing

**Comment by Powersource** [comment]
Ok managed to test locally now and seems to work :)

**Comment by cobraz** [comment]
Did you manage to get them fixed? The build seems to fail, so maybe you can add your fix? :)

**Comment by Powersource** [comment]
No I don't have a fix for the CI stuff. My local fix was to build the package and then remove the node_modules so when I `npm link`ed npm wouldn't get confused about the react versions. Might take a look at the CI stuff though.

**Comment by Powersource** [comment]
Basically just needed to update the tests so that also UserManager.events also is mocked.

**Comment by Powersource** [comment]
Anything I can help with to help get this merged?

File: src/AuthContext.tsx [hunk 1]
```diff
@@ -104,6 +104,18 @@ export const AuthProvider: FC<AuthProviderProps> = ({
...
+    // for refreshing react state when new state is available in e.g. session storage
+    const updateUserData = async () => {
+      const user = await userManager.getUser();
+      setUserData(user);
+    }
+
+    userManager.events.addUserLoaded(updateUserData);
+
+    return () => userManager.events.removeUserLoaded(updateUserData);
+  }, [userManager])
```
Comment (cristian-eriomenco) [inline src/AuthContext.tsx:117]: @Powersource  Could you explain please why would userManager change if it has an execution path where its ignoring props? line #46

Comment (Powersource) [reply src/AuthContext.tsx:117]: I'm no react pro so not gonna defend this too much. I believe https://www.npmjs.com/package/eslint-plugin-react-hooks told me to change that line to this.

But the user could switch out the usermanager they're passing in as a prop, line 36. And then we'd need to re-run this Effect to get it to apply the event handler on the right usermanager.

**Review by ryancole — COMMENTED** [review]
I don't think the `UserManager` is supposed to change, which would mean a React Effect on it wouldn't really make sense to me.

If there is an issue with the underlying oidc-client library not refreshing tokens, then perhaps this would be better as an issue on the repo for that client? [That library](https://github.com/IdentityModel/oidc-client-js) is what handles the refreshing, as far as I can tell.

**Comment by Powersource** [comment]
@ryancole 

Does it maybe make sense to still have the Effect to still have the `removeUserLoaded` if/when the AuthProvider stops being rendered but maybe due to the way and app's code is structured the UserManager object still survives?

The problem isn't that oidc-client fails to refresh. It does that correctly. The problem is that this library doesn't reload the refreshed state into the `userData` state hook.

**Comment by ryancole** [comment]
@Powersource Ah, OK I see now. You're right. The user data state needs to be updated when oidc-client refreshes it. That effect looks fine to me, but perhaps the dependency on `userManager` is not needed? Perhaps just an empty dependency array so that it sets up the event handler on first load?

In the scenario in which we use this library, our `AuthProvider` is at the top of our React component tree and persists through the lifetime of the app. Our UserManager never changes.

In your scenario, is UserManager changing over time?

**Comment by Powersource** [comment]
Removed userManager from the dep array now :)

**Comment by ryancole** [comment]
Looks good to me. /cc @cobraz

**Review by cobraz — APPROVED** [review]

**Comment by Powersource** [comment]
Thanks!

**Pull request merged**